zookeeper-server/src/main/java/org/apache/zookeeper/cli/WatchCommand.java

```diff
@@ -0,0 +1,129 @@
+    }
+
+    @Override
+    public CliCommand parse(String[] cmdArgs) throws CliParseException {
+        Parser parser = new PosixParser();
+        try {
+            cl = parser.parse(options, cmdArgs);
+        } catch (ParseException ex) {
+            throw new CliParseException(ex);
+        }
+        args = cl.getArgs();
+        if (args.length < 2) {
+            throw new CliParseException(getUsageStr());
+        }
+
+        return this;
+    }
+
+    @Override
+    public boolean exec() throws CliException {
+        try {
+            String path = args[1];
+            CountDownLatch latch = new CountDownLatch(1);
+            SimpleCLIWatcher cliWatcher = new SimpleCLIWatcher(latch);
+            if (cl.hasOption("d") || cl.getOptions().length == 0) {
+                zk.getData(path, cliWatcher, null);
+                latch.await();
+                WatchedEvent watchedEvent = cliWatcher.getEvent();
+                printWatchedEvent(watchedEvent);
+                if (watchedEvent.getType().equals(Watcher.Event.EventType.NodeDataChanged)) {
+                    byte[] newData = zk.getData(path, false, null);
+                    out.println("new data:" + new String(newData));
+                }
+            } else if (cl.hasOption("c")) {
+                zk.getChildren(path, cliWatcher);
+                latch.await();
+                WatchedEvent watchedEvent = cliWatcher.getEvent();
+                printWatchedEvent(watchedEvent);
+                if ((watchedEvent.getType().equals(Watcher.Event.EventType.NodeChildrenChanged))) {
+                    List<String> newChildList = zk.getChildren(path, false);
+                    out.println("new child list:" + newChildList.toString());
+                }
+            } else if (cl.hasOption("e")) {
+                zk.exists(path, cliWatcher);
+                latch.await();
+                WatchedEvent watchedEvent = cliWatcher.getEvent();
+                printWatchedEvent(watchedEvent);
+            }
+        } catch (KeeperException | InterruptedException ex) {
+            throw new CliWrapperException(ex);
+        }
+
+        return false;
+    }
+
+    private void printWatchedEvent(WatchedEvent watchedEvent) {
+        out.println("WatchedEvent state:" + watchedEvent.getState());
+        out.println("type:" + watchedEvent.getType());
+        out.println("path:" + watchedEvent.getPath());
+    }
+
+    private static class SimpleCLIWatcher implements Watcher {
+        private CountDownLatch latch;
+        private WatchedEvent event;
+
+        public SimpleCLIWatcher(CountDownLatch latch) {
+            this.latch = latch;
+        }
+
+        public WatchedEvent getEvent() {
+            return event;
+        }
+
+        public void process(WatchedEvent e) {
+            if (e.getType() == Event.EventType.None) {
+                return;
+            }
+            if (e.getState() == Event.KeeperState.SyncConnected) {
```

Comment:
Does this capture every state that might be encountered? Would it be possible to receive a Disconnected or an AuthFailed event and is there something that might gracefully handle them?

The code for this change looks good to me outside of these questions.

Reply:
@enixon  Yes,agreed!